interfaces/builtin/optical_drive.go

```diff
@@ -28,27 +37,73 @@ const opticalDriveBaseDeclarationSlots = `
 
-const opticalDriveConnectedPlugAppArmor = `
-# Allow read access to optical drives
-/dev/sr[0-9]* r,
-/dev/scd[0-9]* r,
-@{PROC}/sys/dev/cdrom/info r,
-/run/udev/data/b11:[0-9]* r,
-`
+// opticalDriveInterface is the type for optical drive interfaces.
+type opticalDriveInterface struct {
+	commonInterface
+}
+
+// BeforePrepareSlot checks and possibly modifies a slot.
+// Valid "optical-drive" slots may contain the attribute "write".
+// If defined, the attribute "write" must be either "true" or "false".
+func (iface *opticalDriveInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	write, ok := plug.Attrs["write"].(bool)
+	if !ok {
+		if plug.Attrs == nil {
+			plug.Attrs = make(map[string]interface{})
+		}
+		write = false
+		plug.Attrs["write"] = false
+	}
+
+	if write != true && write != false {
+		return fmt.Errorf("optical-drive write attribute must be either undefined, true, or false.")
+	}
+
+	return nil
+}
+
+func (iface *opticalDriveInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	spec.AddSnippet("/run/udev/data/b11:[0-9]* r,")
+	spec.AddSnippet("@{PROC}/sys/dev/cdrom/info r,")
+
+	var write bool
+	if err := plug.Attr("write", &write); err == nil && write == true {
+		// Allow read and write access to optical drive block devices
+		spec.AddSnippet("/dev/sr[0-9]* rw,")
+		spec.AddSnippet("/dev/scd[0-9]* rw,")
+		return nil
+	}
+
+	// Allow readonly access to optical drive block devices
+	spec.AddSnippet("/dev/sr[0-9]* r,")
+	spec.AddSnippet("/dev/scd[0-9]* r,")
+	return nil
+}
+
+func (iface *opticalDriveInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	spec.TagDevice(`KERNEL=="sr[0-9]*"`)
+	spec.TagDevice(`KERNEL=="scd[0-9]*"`)
+	return nil
+}
+
+func (iface *opticalDriveInterface) AutoConnect(plug *interfaces.Plug, _ *interfaces.Slot) bool {
+	var write bool
+	// Prevent auto connection when write is requested
+	if err := plug.Attr("write", &write); err == nil && write == true {
+		return false
+	}
 
-var opticalDriveConnectedPlugUDev = []string{
-	`KERNEL=="sr[0-9]*"`,
-	`KERNEL=="scd[0-9]*"`,
+	// Allow auto connection when write is not requested
+	return true
 }
 
 func init() {
-	registerIface(&commonInterface{
-		name:                  "optical-drive",
-		summary:               opticalDriveSummary,
-		implicitOnClassic:     true,
-		baseDeclarationSlots:  opticalDriveBaseDeclarationSlots,
-		connectedPlugAppArmor: opticalDriveConnectedPlugAppArmor,
-		connectedPlugUDev:     opticalDriveConnectedPlugUDev,
-		reservedForOS:         true,
-	})
+	registerIface(&opticalDriveInterface{commonInterface: commonInterface{
+		name:                 "optical-drive",
+		summary:              opticalDriveSummary,
+		implicitOnCore:       false,
+		implicitOnClassic:    true,
+		baseDeclarationSlots: opticalDriveBaseDeclarationSlots,
+		reservedForOS:        true,
+	}})
 }
```

Comment:
It might be hard to process all of the above. I suggest the file be something along the lines of (untested):
```
...
const opticalDriveSummary = `allows read by default and optionally write access to optical drives`

const opticalDriveBaseDeclarationSlots = `
  optical-drive:
    allow-installation:
      slot-snap-type:
        - core
    deny-connection:
      plug-attributes:
        write: true
    deny-auto-connection:
       plug-attributes:
         write: true
`

const opticalDriveConnectedPlugAppArmor = `
# Allow read access to optical drives
/dev/sr[0-9]* r,
/dev/scd[0-9]* r,
@{PROC}/sys/dev/cdrom/info r,
/run/udev/data/b11:[0-9]* r,
`

var opticalDriveConnectedPlugUDev = []string{
	`KERNEL=="sr[0-9]*"`,
	`KERNEL=="scd[0-9]*"`,
}

func (iface *opticalDriveInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
       // It's fine if 'write' isn't specified, but if it is, it needs to be bool
       if w, ok := plug.Attrs["write"]; ok {
               _, ok = w.(bool)
               if !ok {
                       return fmt.Errorf(`optical-drive plug requires "write" be bool`)
               }
       }

       return nil
}

func (iface *opticalDriveInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
       var write bool
       _ = plug.Attr("write", &write)

       // Add the common readonly policy
       spec.AddSnippet(opticalDriveConnectedPlugAppArmor)

       // 'write: true' grants write access to the devices
       if write {
               spec.AddSnippet("# Allow write access to optical drives")
               spec.AddSnippet("/dev/sr[0-9]* w,")
               spec.AddSnippet("/dev/scd[0-9]* w,")
       }
       return nil
}

func init() {
	registerIface(&opticalDriveInterface{commonInterface: commonInterface{
		name:                 "optical-drive",
		summary:              opticalDriveSummary,
		implicitOnCore:       false,
		implicitOnClassic:    true,
		baseDeclarationSlots: opticalDriveBaseDeclarationSlots,
		connectedPlugUDev:    opticalDriveConnectedPlugUDev,
		reservedForOS:        true,
	}})
}
```